compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/JvmLower.kt

```diff
@@ -184,15 +184,11 @@ private val syntheticAccessorPhase = makeIrFilePhase(
     prerequisite = setOf(objectClassPhase, staticDefaultFunctionPhase, interfacePhase)
 )
 
-private val mainMethodGenerationPhase = makeIrFilePhase(
-    ::MainMethodGenerationLowering,
-    name = "MainMethodGeneration",
-    description = "Identify parameterless main methods and generate bridge main-methods"
-)
-
 @Suppress("Reformat")
 private val jvmFilePhases =
+        jvmOverloadsAnnotationPhase then
```

Comment:
Can we move `mainMethodGenerationPhase` at the end instead? We would probably like to have a common prefix of lowerings across all backends some day in the future, so it seems useful to minimize the number of platform-specific lowerings in the beginning

Reply:
Indeed we can! O.o

I can't recall what prompted me to place it at the front, but I do recall discussing that it would be simpler to do during code-gen when things like `@JvmName` have been dealt with. Essentially parameterless-main is just a patch on the bytecode: do we have a signature `public static void main ()V` but no `public static void main ([LString)V`, add the latter, without regards for annotations, projects, type parameters etc.

Reply:
Ideally we'd have lowerings that remove all these specifics and then operate on the IR that is as close to the bytecode as possible. Modifying the IR is far easier than modifying the bytecode anyway. But I don't know exactly why we don't for example handle `@JvmName` annotations in a lowering. I think this could get rid of JvmName checks in the `mainMethodGenerationPhase`, as well as in `MethodSignatureMapper`. But probably this could introduce a whole set of new problems, so I don't encourage this approach either.

Reply:
`JvmNamePhase` sounds like a small experiment for a rainy day :) I'll add it to my list regardless. Do you at the moment have a sense of where the "common prefix" ends in `JvmLower.kt`?

Reply:
You can get a rough idea at where the common prefix is supposed to end in `JsLoweringPhases.kt` (see the "Common prefix ends" comment), but that's also not set in stone. Note that we're not looking into extracting and maintaining the common prefix in JVM IR right now, it's more of a goal for the future.